src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/net/WireHopperTest.java

```diff
@@ -0,0 +1,179 @@
+        // verify resilience to null executor
+        GenericFlowUnit flowUnit = new SymptomFlowUnit(TIMESTAMP);
+        DataMsg msg = new DataMsg(NODE1, Lists.newArrayList(NODE2), Collections.singletonList(flowUnit));
+        uut.sendData(msg);
+
+        clientExecutor.set(executorService);
+        Assert.assertEquals(0L, nodeStateManager.getLastReceivedTimestamp(NODE1, LOCALHOST));
+        // setup downstream subscribers
+        subscriptionManager.setCurrentLocus(LOCUS);
+        subscriptionManager.addSubscriber(NODE1, LOCALHOST, LOCUS);
+        // verify sendData works
+        ClusterDetailsEventProcessor.setNodesDetails(Lists.newArrayList(
+                new ClusterDetailsEventProcessor.NodeDetails(NODE1, LOCALHOST, "", false),
+                new ClusterDetailsEventProcessor.NodeDetails(NODE2, LOCALHOST, "", false)
+        ));
+        uut.sendData(msg);
+        Thread.sleep(2000L);
+        Assert.assertNotEquals(0L, nodeStateManager.getLastReceivedTimestamp(NODE1, LOCALHOST));
+        List<FlowUnitMessage> receivedMags = receivedFlowUnitStore.drainNode(NODE1);
+        Assert.assertEquals(1L, receivedMags.size());
+
+        // verify resilience to RejectedExecutionException
+        clientExecutor.set(rejectingExecutor);
+        uut.sendData(msg);
+    }
+
+    @Test
+    public void testReadFromWire() throws InterruptedException {
+        netServer.setSubscribeHandler(new SubscribeServerHandler(subscriptionManager, serverExecutor));
+        // Setup mock object responses
+        Node<MetricFlowUnit> node = new Heap_Used(EVAL_INTERVAL_S);
+        node.addTag(RcaConsts.RcaTagConstants.TAG_LOCUS, RcaConsts.RcaTagConstants.LOCUS_DATA_MASTER_NODE);
+        // Verify resilience to null executor
+        uut.readFromWire(node);
+        // Execute test method and verify return value
+        clientExecutor.set(executorService);
+        ClusterDetailsEventProcessor.setNodesDetails(Collections.singletonList(
+                new ClusterDetailsEventProcessor.NodeDetails(
+                        node.name(), LOCALHOST, RcaConsts.RcaTagConstants.LOCUS_DATA_NODE, false)));
+        subscriptionManager.setCurrentLocus(RcaConsts.RcaTagConstants.LOCUS_DATA_NODE);
+        subscriptionManager.addPublisher(node.name(), LOCALHOST);
+        subscriptionManager.addPublisher(node.name(), HOST_NOT_IN_CLUSTER);
+        nodeStateManager.updateReceiveTime(LOCALHOST, node.name(), 1L);
+        FlowUnitMessage msg = FlowUnitMessage.newBuilder().setGraphNode(node.name()).build();
+        ImmutableList<FlowUnitMessage> msgList = ImmutableList.<FlowUnitMessage>builder().add(msg).build();
+        receivedFlowUnitStore.enqueue(node.name(), msg);
+        List<FlowUnitMessage> actualMsgList = uut.readFromWire(node);
+        Assert.assertEquals(msgList, actualMsgList);
+        // Verify expected interactions with the subscription manager
+        Thread.sleep(3000L);
```

Comment:
How did we arrive at this number ? I have seen arbitrary sleeps make test cases flaky. And sometimes the tests take too long to run because there are multiple tests doing this. It might be a good idea to check for the expected value in a loop for some max timeouts where the sleeps between each iteration is in milliseconds. That way we don't wait more than we have to. and the timeout can be much longer than 3 seconds as we expect to break out of them early anyways

Reply:
I've modified everything to be a uniform 1 second. Testing the mean on my local machine would likely differ from the mean on the CI runner, so I feel that 1s is a good uniform compromise.

Reply:
Added the WaitFor class which is a basic reusable waiting poller